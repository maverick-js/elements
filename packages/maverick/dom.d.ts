export * from './dist/prod/dom.js';
