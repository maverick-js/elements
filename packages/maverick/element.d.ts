export * from './dist/prod/element.js';
