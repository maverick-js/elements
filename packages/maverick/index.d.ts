export * from './dist/prod/runtime.js';
