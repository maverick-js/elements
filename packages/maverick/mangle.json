{
  "_$state": "h",
  "_addHooks": "t",
  "_attach": "x",
  "_attachAttrs": "q",
  "_attachCallbacks": "m",
  "_attachScope": "f",
  "_attachStyles": "r",
  "_attached": "Q",
  "_attrs": "b",
  "_callbacks": "J",
  "_component": "e",
  "_connect": "y",
  "_connectCallbacks": "j",
  "_connectId": "E",
  "_connectScope": "g",
  "_context": "F",
  "_converter": "B",
  "_currentRenderer": "X",
  "_currentRenderer2": "Y",
  "_currentValue": "M",
  "_currentValue2": "N",
  "_defaultValue": "V",
  "_descriptors": "A",
  "_destroy": "p",
  "_destroyCallbacks": "n",
  "_destroyed": "o",
  "_detach": "s",
  "_disconnect": "z",
  "_el": "a",
  "_globalName": "W",
  "_init": "S",
  "_owner": "G",
  "_payload": "R",
  "_prop": "C",
  "_props": "i",
  "_provide": "T",
  "_refChangeId": "K",
  "_result": "I",
  "_scope": "d",
  "_self": "U",
  "_setAttr": "u",
  "_setStyle": "v",
  "_setup": "w",
  "_setupCallbacks": "l",
  "_source": "L",
  "_state": "k",
  "_status": "D",
  "_store": "H",
  "_styles": "c",
  "_threadCount": "O",
  "_tokens": "Z",
  "_updatedFibers": "P",
  "_domCallbacks": "_",
  "_prevProps": "aa",
  "_onDOMEvent": "ba",
  "_forwardedRef": "da",
  "_listening": "ca",
  "_destroyId": "ea"
}