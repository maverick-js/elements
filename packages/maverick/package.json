{
  "name": "maverick.js",
  "description": "Build functional, reactive, type-safe, lightweight, and fast UI libraries.",
  "license": "MIT",
  "version": "0.32.3",
  "type": "module",
  "types": "dist/types/runtime/index.d.ts",
  "main": "dist/server/runtime.cjs",
  "module": "dist/prod/runtime.js",
  "sideEffects": false,
  "files": [
    "dist/",
    "*.d.ts"
  ],
  "engines": {
    "node": ">=16"
  },
  "contributors": [
    "Rahim Alwer <rahim.alwer@gmail.com>"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/maverick-js/maverick.git",
    "directory": "packages/maverick"
  },
  "bugs": {
    "url": "https://github.com/maverick-js/maverick/issues"
  },
  "scripts": {
    "dev": "pnpm run build --watch",
    "build": "rimraf dist && tsup && pnpm types",
    "types": "tsc -p tsconfig.build.json",
    "test": "pnpm test:client && pnpm test:server",
    "format": "prettier src --write --loglevel warn",
    "test:client": "vitest --run",
    "test:client:watch": "vitest --watch",
    "test:server": "SERVER=true vitest --run",
    "test:server:watch": "SERVER=true vitest --watch"
  },
  "dependencies": {
    "@maverick-js/signals": "^5.9.1",
    "type-fest": "^2.17.0"
  },
  "devDependencies": {
    "@edge-runtime/vm": "^2.0.0",
    "@maverick-js/compiler": "workspace:*",
    "@types/node": "^18.0.3",
    "@types/react": "^18.0.24",
    "@types/react-dom": "^18.0.8",
    "benchmark": "^2.1.4",
    "jsdom": "^20.0.0",
    "node-fetch": "^3.2.10",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rimraf": "^3.0.2",
    "tsup": "^6.2.3",
    "typescript": "^4.9.0",
    "vite": "^3.2.3",
    "vitest": "^0.25.1"
  },
  "exports": {
    ".": {
      "import": {
        "test": "./dist/dev/runtime.js",
        "node": "./dist/server/runtime.js",
        "deno": "./dist/server/runtime.js",
        "bun": "./dist/server/runtime.js",
        "development": "./dist/dev/runtime.js",
        "default": "./dist/prod/runtime.js"
      },
      "require": "./dist/server/runtime.cjs"
    },
    "./dom": {
      "import": {
        "test": "./dist/dev/dom.js",
        "node": "./dist/server/dom.js",
        "deno": "./dist/server/dom.js",
        "bun": "./dist/server/dom.js",
        "development": "./dist/dev/dom.js",
        "default": "./dist/prod/dom.js"
      },
      "require": "./dist/server/dom.cjs"
    },
    "./ssr": {
      "import": {
        "test": "./dist/dev/ssr.js",
        "node": "./dist/server/ssr.js",
        "deno": "./dist/server/ssr.js",
        "bun": "./dist/server/ssr.js",
        "development": "./dist/dev/ssr.js",
        "default": "./dist/prod/ssr.js"
      },
      "require": "./dist/server/ssr.cjs"
    },
    "./element": {
      "import": {
        "test": "./dist/dev/element.js",
        "node": "./dist/server/element.js",
        "deno": "./dist/server/element.js",
        "bun": "./dist/server/element.js",
        "development": "./dist/dev/element.js",
        "default": "./dist/prod/element.js"
      },
      "require": "./dist/server/element.cjs"
    },
    "./react": {
      "import": {
        "test": "./dist/dev/react.js",
        "node": "./dist/server/react.js",
        "deno": "./dist/server/react.js",
        "bun": "./dist/server/react.js",
        "development": "./dist/dev/react.js",
        "default": "./dist/prod/react.js"
      },
      "require": "./dist/server/react.cjs"
    },
    "./std": {
      "import": {
        "test": "./dist/dev/std.js",
        "node": "./dist/server/std.js",
        "deno": "./dist/server/std.js",
        "bun": "./dist/server/std.js",
        "development": "./dist/dev/std.js",
        "default": "./dist/prod/std.js"
      },
      "require": "./dist/server/std.cjs"
    },
    "./package.json": "./package.json"
  },
  "publishConfig": {
    "access": "public"
  }
}
