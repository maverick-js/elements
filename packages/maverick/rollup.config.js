import { nodeResolve } from '@rollup/plugin-node-resolve';
import { build } from 'esbuild';
import { globbySync } from 'globby';
import fs from 'node:fs/promises';
import { defineConfig } from 'rollup';
import dts from 'rollup-plugin-dts';
import esbuildPlugin from 'rollup-plugin-esbuild';

/** @type {Record<string, string | false>} */
const MANGLE_CACHE = await buildMangleCache();

export default defineConfig([
  define({ type: 'dev' }),
  define({ type: 'prod' }),
  define({ type: 'server' }),
  define({ type: 'rsc-dev' }),
  define({ type: 'rsc-prod' }),
  defineTypes(),
]);

/** @returns {import('rollup').RollupOptions} */
function defineTypes() {
  return {
    input: {
      index: 'types/core/index.d.ts',
      element: 'types/element/index.d.ts',
      react: 'types/react/index.d.ts',
      std: 'types/std/index.d.ts',
    },
    output: {
      dir: '.',
      chunkFileNames: 'dist/types/maverick-[hash].d.ts',
      compact: false,
      minifyInternalExports: false,
    },
    external: ['react'],
    plugins: [
      dts({ respectExternal: true }),
      {
        name: 'cleanup',
        async closeBundle() {
          await fs.rm('types', { recursive: true });
        },
      },
    ],
  };
}

/**
 * @typedef {{
 *   type: 'dev' | 'prod' | 'server' | 'rsc-dev' | 'rsc-prod';
 * }} BundleOptions
 */

/**
 * @param {BundleOptions}
 * @returns {import('rollup').RollupOptions}
 */
function define({ type = 'dev' }) {
  const isDev = type === 'dev' || type === 'rsc-dev',
    isRSC = type.startsWith('rsc');

  return {
    input: isRSC
      ? { [type.replace('rsc-', '')]: 'src/react/rsc.ts' }
      : {
          index: 'src/core/index.ts',
          element: 'src/element/index.ts',
          react: 'src/react/index.ts',
          std: 'src/std/index.ts',
        },
    treeshake: true,
    preserveEntrySignatures: 'allow-extension',
    external: ['react'],
    output: {
      format: 'esm',
      dir: isRSC ? `dist/rsc` : `dist/${type}`,
      chunkFileNames: `chunks/maverick-[hash].js`,
      compact: false,
      minifyInternalExports: false,
    },
    plugins: [
      nodeResolve({
        exportConditions: isDev
          ? ['development', 'production', 'default']
          : ['production', 'default'],
      }),
      esbuildPlugin({
        target: type === 'server' ? 'node18' : 'esnext',
        platform: type === 'server' ? 'node' : 'browser',
        tsconfig: 'tsconfig.build.json',
        minify: false,
        mangleProps: !isDev ? /^_/ : undefined,
        reserveProps: !isDev ? /^__/ : undefined,
        mangleCache: !isDev ? MANGLE_CACHE : undefined,
        banner: isRSC ? 'import { IS_SERVER } from "@virtual/env";\n' : '',
        define: {
          __DEV__: isDev ? 'true' : 'false',
          __SERVER__: isRSC ? 'IS_SERVER' : type === 'server' ? 'true' : 'false',
          __TEST__: 'false',
        },
      }),
      {
        name: 'env',
        resolveId(id) {
          if (id === '@virtual/env') return id;
        },
        load(id) {
          if (id === '@virtual/env') {
            return `export const IS_SERVER = typeof document === 'undefined';`;
          }
        },
      },
    ],
  };
}

async function buildMangleCache() {
  let cache = JSON.parse(await fs.readFile('mangle.json', 'utf-8'));

  const result = await build({
    entryPoints: globbySync('src/**'),
    target: 'esnext',
    bundle: true,
    minify: false,
    mangleProps: /^_/,
    reserveProps: /^__/,
    mangleCache: cache,
    write: false,
    outdir: 'dist-esbuild',
  });

  cache = {
    ...cache,
    ...result.mangleCache,
  };

  await fs.writeFile('mangle.json', JSON.stringify(cache, null, 2));

  return cache;
}
