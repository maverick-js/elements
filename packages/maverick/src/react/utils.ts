export function setRef(ref: React.Ref<unknown>, value: Element | null) {
  if (typeof ref === 'function') {
    (ref as (e: Element | null) => void)(value);
  } else {
    (ref as { current: Element | null }).current = value;
  }
}
