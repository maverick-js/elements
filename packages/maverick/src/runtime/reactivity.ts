import { effect as $effect, scope } from '@maverick-js/signals';

import { noop } from '../std/unit';

/**
 * Invokes the given function each time any of the signals that are read inside are updated
 * (i.e., their value changes). The effect is immediately invoked on initialization.
 *
 * @see {@link https://github.com/maverick-js/signals#effect}
 */
export const effect = (__SERVER__ ? noop : $effect) as typeof $effect;

/**
 * Creates a runner that executes functions in the current lexical scope.
 */
export function createScopedRunner() {
  let currentFn: (() => any) | undefined,
    run = scope(() => currentFn?.());
  return function runWithScope<T>(fn: () => T) {
    currentFn = fn;
    const result = run();
    currentFn = undefined;
    return result;
  };
}

export * from '@maverick-js/signals';
export * from './store';
