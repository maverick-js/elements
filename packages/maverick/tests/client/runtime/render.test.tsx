import { observable, render, tick } from 'maverick.js';

import { Element, Fragment } from './fixtures/primitives';
import { InputField } from './fixtures/reactivity';
import { element } from './utils';

it('should render element', () => {
  const root = element('root');

  render(() => <Element />, { target: root });

  expect(root).toMatchInlineSnapshot(`
    <root>
      <div />
    </root>
  `);

  expect(root.firstElementChild).toBeInstanceOf(HTMLDivElement);
  expect(<Element />).toBeInstanceOf(HTMLDivElement);
});

it('should render fragment', () => {
  const root = element('root');

  render(() => <Fragment />, { target: root });

  expect(root).toMatchInlineSnapshot(`
    <root>
      <div />
      <div />
      <div />
    </root>
  `);

  expect(Array.isArray(<Fragment />)).toBeTruthy();
  expect(((<Fragment />) as any[])[0]).toBeInstanceOf(HTMLDivElement);
});

it('should render components', () => {
  function Component(props) {
    return props.$children;
  }

  function ChildComponent(props) {
    return <div>{props.$children}</div>;
  }

  const root = element('root');

  render(
    () => (
      <Component>
        Text
        <div>Text</div>
        <ChildComponent>Child Text</ChildComponent>
        <ChildComponent>Child Text</ChildComponent>
        <div>Text</div>
        <ChildComponent />
      </Component>
    ),
    { target: root },
  );

  expect(root).toMatchInlineSnapshot(`
    <root>
      Text
      <div>
        Text
      </div>
      <div>
        <!--$-->
        Child Text
        <!--/$-->
      </div>
      <div>
        <!--$-->
        Child Text
        <!--/$-->
      </div>
      <div>
        Text
      </div>
      <div>
        <!--$-->
        <!--/$-->
      </div>
    </root>
  `);

  expect(<Fragment />).toBeInstanceOf(Array);
});

it('should be reactive', async () => {
  const root = element('root');
  let next!: () => void;

  const input = <InputField next={(n) => (next = n)} />;

  render(() => input, { target: root });
  await tick();

  expect(root).toMatchInlineSnapshot(`
    <root>
      <div>
        <span>
          Count is 
          <!--$-->
          1
          <!--/$-->
        </span>
        <!--$-->
        <input
          type="number"
        />
      </div>
    </root>
  `);

  const getValueTextNode = () => root.querySelector('span')!.childNodes[2];

  const valueText = getValueTextNode();
  expect(valueText).toBeInstanceOf(Text);

  const inputElement = root.querySelector('input');
  expect(inputElement).toBeInstanceOf(HTMLInputElement);

  next();
  await tick();

  expect(root).toMatchInlineSnapshot(`
    <root>
      <div>
        <span>
          Count is 
          <!--$-->
          2
          <!--/$-->
        </span>
        <!--$-->
        <input
          type="number"
        />
      </div>
    </root>
  `);

  // it should re-use existing node.
  expect(getValueTextNode()).toBe(valueText);

  inputElement!.dispatchEvent(new CustomEvent('next'));
  await tick();

  expect(root).toMatchInlineSnapshot(`
    <root>
      <div>
        <span>
          Count is 
          <!--$-->
          3
          <!--/$-->
        </span>
        <!--$-->
        <input
          type="number"
        />
      </div>
    </root>
  `);
});

it('should render observable component', async () => {
  function Component(props) {
    return props.$children;
  }

  const $count = observable(1);

  const $component = observable(
    <Component>
      <span>{$count()}</span>
      <span>{$count()}</span>
    </Component>,
  );

  const root = element('root');
  render(() => $component, { target: root });

  expect(root).toMatchInlineSnapshot(`
    <root>
      <!--$$-->
      <span>
        <!--$-->
        1
        <!--/$-->
      </span>
      <span>
        <!--$-->
        1
        <!--/$-->
      </span>
      <!--/$-->
    </root>
  `);

  $component.set(
    <Component>
      <span>Foo</span>
      <span>Bar</span>
    </Component>,
  );

  await tick();

  expect(root).toMatchInlineSnapshot(`
    <root>
      <!--$$-->
      <span>
        Foo
      </span>
      <span>
        Bar
      </span>
      <!--/$-->
    </root>
  `);
});
