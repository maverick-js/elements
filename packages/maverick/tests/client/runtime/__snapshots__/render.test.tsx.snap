// Vitest Snapshot v1

exports[`should be reactive 1`] = `
<root>
  <div>
    <span>
      Count is 
      1
      <!--$-->
    </span>
    <!--$-->
    <input
      type="number"
    />
  </div>
</root>
`;

exports[`should be reactive 2`] = `
<root>
  <div>
    <span>
      Count is 
      2
      <!--$-->
    </span>
    <!--$-->
    <input
      type="number"
    />
  </div>
</root>
`;

exports[`should be reactive 3`] = `
<root>
  <div>
    <span>
      Count is 
      3
      <!--$-->
    </span>
    <!--$-->
    <input
      type="number"
    />
  </div>
</root>
`;
